File: Java/55_Jump_Game.java
class Solution {
    int [] getTo; //global array to store results
    public boolean canJump(int[] nums) {
        getTo = new int[nums.length + 1];
        
        return jumpHelper(nums,0);
    }
    public boolean jumpHelper(int [] nums, int index) {
        //if we reached end we made it so we can return
        if(index >= nums.length - 1 || getTo[index] == 1) {
            return true;
        } else if(getTo[index] == -1) {
            //if we have found the index cannot be reached we return false
            return false;
        }
        //go through all of the jumps possible and check if we can make it
        for(int i = nums[index]; i >= 1;i--) {
            if(jumpHelper(nums,index + i)) {
                //if true we store result in array and return
                getTo[index] = 1;
                return true;
            }
        }
        //otherwise we store the result and return false
        getTo[index] = -1;
        return false;
    }
}

File: C++/55_Jump_Game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int jumps_left = 0;
        for(int i = 0; i < nums.size();i++) {
            if(jumps_left < 0) {
                return false;
            } 
            /* The key idea here is that there are TWO cases:
            if we have more jumps then the current index, we just keep using those, but
            if we can gain more, we should just do that by starting from the current idx
            */ 
            jumps_left = max(jumps_left, nums[i]);
            jumps_left--;
        }
        return true;
    }
};
